src/main/java/bio/terra/service/job/JobService.java

```diff
@@ -323,43 +330,48 @@ public List<JobModel> enumerateJobs(
   private List<FlightState> getAccessibleFlights(
       int offset, int limit, FlightFilter filter, AuthenticatedUserRequest userReq)
       throws InterruptedException {
-    int start = 0;
-    int end = offset + limit;
+
+    String nextPageToken = null;
     HashMap<String, List<String>> roleMap = new HashMap<>();
     ArrayList<FlightState> flightStateList = new ArrayList<>();
-    while (flightStateList.size() < offset + limit) {
-      List<FlightState> filterResults = stairway.getFlights(start, end, filter);
-      if (filterResults.size() == 0) {
+    int numTimesQueried = 0;
+    while (flightStateList.size() < offset + limit && numTimesQueried < MAX_FLIGHT_SEARCH_QUERIES) {
+      logger.info("Getting flight batch {} with npt {}", numTimesQueried, nextPageToken);
+      FlightEnumeration filterResults =
+          stairway.getFlights(nextPageToken, FLIGHT_SEARCH_BATCH_SIZE, filter);
+      if (filterResults.getFlightStateList().size() == 0) {
         break;
       }
-      filterResults.forEach(
-          flightState -> {
-            if (userLaunchedFlight(flightState, userReq)) {
-              flightStateList.add(flightState);
-            } else {
-              FlightMap inputParameters = flightState.getInputParameters();
-              IamResourceType resourceType =
-                  inputParameters.get(
-                      JobMapKeys.IAM_RESOURCE_TYPE.getKeyName(), IamResourceType.class);
-              String resourceId =
-                  inputParameters.get(JobMapKeys.IAM_RESOURCE_ID.getKeyName(), String.class);
-              IamAction action =
-                  inputParameters.get(JobMapKeys.IAM_ACTION.getKeyName(), IamAction.class);
-              if (resourceType != null & resourceId != null && action != null) {
-                String key = String.format("%s-%s", resourceType, resourceId);
-                List<String> userRoles = roleMap.get(key);
-                if (userRoles == null) {
-                  userRoles = samService.listActions(userReq, resourceType, resourceId);
-                  roleMap.put(key, userRoles);
-                }
-                if (userRoles.contains(action.toString())) {
+      filterResults
+          .getFlightStateList()
+          .forEach(
```

Comment:
With a `FLIGHT_SEARCH_BATCH_SIZE` of 1,000 we are doing more work than necessary when given a lower limit, since we have to go through each flight state and determine whether it's indirectly accessible to the caller.  Not sure if it makes that big a difference performance-wise, though.

What do you think about breaking out of the loop when you've hit the limit?

Reply:
Oh that's a great point.  Will update